print_nanny_webapp/subscriptions/views.py

```diff
@@ -0,0 +1,172 @@
+    )
+    return JsonResponse({"session_id": session.id}, status=200)
+
+@webhooks.handler("customer.subscription.trial_will_end")
+def subscriptions_trial_will_end(event):
+    logger.info(f"Sending email to user {event.customer.email} that their trial is ending")
+    user = User.objects.get(email=event.customer.email)
+
+    merge_data = {
+        "FIRST_NAME": user.first_name or "Maker",
+    }
+
+    text_body = render_to_string("email/subscriptions_trial_ending_body.txt", merge_data)
+    # html_body = render_to_string("email/subscriptions_trial_ending_body.html", merge_data)
+    subject = render_to_string("email/subscriptions_trial_ending_subject.txt", merge_data)
+
+    message = AnymailMessage(
+        subject=subject,
+        body=text_body,
+        to=[event.customer.email],
+    )
+    # message.attach_alternative(html_body, "text/html")
+    message.send()
+
+@webhooks.handler("subscription_schedule.expiring")
+def subscriptions_subscription_expiring(event):
+    logger.info(f"User {event.customer.email} subscription <X> is expiring, sending email")
+    user = User.objects.get(email=event.customer.email)
+
+    merge_data = {
+        "FIRST_NAME": user.first_name or "Maker",
+    }
+
+    text_body = render_to_string("email/subscriptions_subscription_expiring_body.txt", merge_data)
+    # html_body = render_to_string("email/subscriptions_subscription_expiring_body.html", merge_data)
+    subject = render_to_string("email/subscriptions_subscription_expiring_subject.txt", merge_data)
+
+    message = AnymailMessage(
+        subject=subject,
+        body=text_body,
+        to=[event.customer.email],
+    )
+    # message.attach_alternative(html_body, "text/html")
+    message.send()
+
+# @webhooks.handler("charge.failed")
+@webhooks.handler("invoice.payment_failed")
+def subscriptions_payment_failed(event):
+    logger.info(f"User's {event.customer.email} subscription <X> payment <X> failed, sending email")
+    user = User.objects.get(email=event.customer.email)
+
+    merge_data = {
+        "FIRST_NAME": user.first_name or "Maker",
+    }
+
+    text_body = render_to_string("email/subscriptions_payment_failed_body.txt", merge_data)
+    # html_body = render_to_string("email/subscriptions_subscription_expiring_body.html", merge_data)
+    subject = render_to_string("email/subscriptions_payment_failed_subject.txt", merge_data)
+
+    message = AnymailMessage(
+        subject=subject,
+        body=text_body,
+        to=[event.customer.email],
+    )
+    # message.attach_alternative(html_body, "text/html")
+    message.send()
+
+# Payment action required is handled during the payment step inside the view
+# There are both invoice.paid and invoice.payment_succeeded but docs lean on the first
+# https://stripe.com/docs/billing/subscriptions/webhooks#tracking
+# @webhooks.handler("charge.succeeded")
+@webhooks.handler("invoice.paid")
+def subscriptions_invoice_paid(event):
+    # TODO: When a trial was paid, do nothing
```

Comment:
Reading between the lines, does a trial have a separate paid/unpaid status? :thinking: That would make sense, since some companies want to run paid or discounted trials.

Reply:
apparently not. I just searched my emails and I didn't get an invoice paid mail during account creation. Most probably leftover TODO from the `charge.succeeded` webhook